libs/SalesforceSDK/src/com/salesforce/androidsdk/app/SalesforceSDKManager.java

```diff
@@ -848,12 +851,24 @@ private void postPushUnregister(BroadcastReceiver pushReceiver,
     		final ClientManager clientMgr, final boolean showLoginPage,
     		final String refreshToken, final String loginServer,
             final Account account, Activity frontActivity) {
-        try {
-            context.unregisterReceiver(pushReceiver);
-        } catch (Exception e) {
-            SalesforceSDKLogger.e(TAG, "Exception occurred while un-registering", e);
+
+        /*
+         * This method can be triggered from 2 places - either when the push un-registration call
+         * completes, or when the server times out. Since they're on different threads, we don't
+         * know when either of those calls returns. However, the code should be executed only
+         * once. We also need to make sure that the value is reset once both calls come in,
+         * to ensure that push un-registration works in the multi-user scenario.
+         */
+        if (numPostPushUnregisterCalls <= 1) {
+            try {
+                context.unregisterReceiver(pushReceiver);
+            } catch (Exception e) {
+                SalesforceSDKLogger.e(TAG, "Exception occurred while un-registering", e);
+            }
+            removeAccount(clientMgr, showLoginPage, refreshToken, loginServer, account, frontActivity);
+        } else {
+            numPostPushUnregisterCalls = 0;
```

Comment:
I'm not sure it will work in all cases. For instance:
* What if broadcast receiver does `numPostPushUnregisterCalls++`, followed by `postPushUnregister()` which gets to the else and therefore sets `numPostPushUnregisterCalls` back to 0.
At that point, the background thread that was sleeping, does `numPostPushUnregisterCalls++`, followed by `postPushUnregister()`, the `if (numPostPushUnregisterCalls <= 1)` will be entered a second time then.
* What if broadcast receiver does `numPostPushUnregisterCalls++`, and the background thread does its `numPostPushUnregisterCalls++`, before `postPushUnregister()` runs. In that case the `if (numPostPushUnregisterCalls <= 1)` will never be entered.

Reply:
Yeah, it doesn't cover all cases. The foolproof way to do this to solve the single user and multi-user scenarios is to have a `boolean` per `UserAccount`. Only then would we be able to support this use case across multiple users, because resetting the `boolean` at the end is always going to be unpredictable (we don't know how long to wait). If we had 1 `boolean` per `UserAccount`, we could reset it when the user logs back in. Let me take a stab at that.